security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,165 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal/proto"
+	"hash"
+	"sync"
+)
+
+// uint64Size is the size of a uint64 in bytes.
+const uint64Size = 8
+
+const (
+	tls13Key    = "tls13 key"
+	tls13Nonce  = "tls13 iv"
+	tls13Update = "tls13 traffic upd"
+)
+
+type S2AHalfConnection struct {
+	cs            ciphersuite
+	h             func() hash.Hash
```

Comment:
Would it make sense to pass the hash function to the expander, and the expander can own the hash function? Unless I'm missing something, the hash function is only ever used as an argument to the expander.